Swerve subsystem

adityapawar1 opened the pull request:
## Summary

This PR creates the SwerveDrive subsystem and SwerveCommand for WPILib's command based framework

Additions
- SwerveDrive Subsystem
- SwerveDrive Command
- SwerveDrive odometry for autonomous
- Take controller input to controller linear and angular velocity for the swerve subsystem

## Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [x] I have read the [Contributing](Contributing.md) document.

- [ ] If code changes were made then they have been tested.

adityapawar1 committed:
Added swerve subsystem

 Added previous SwerveKinematics in the drive method

adityapawar1 committed:
Added swerve command

adityapawar1 committed:
Created Translations and SwerveDriveKinematics as class variables

adityapawar1 committed:
Added controller helpers

Added dylan's controller helpers from the T-Shirt Shooter repo

adityapawar1 committed:
Updated SwerveCommand.java

Added isFinished and end methods in SwerveCommand.java

akhil-datla committed:
Added SwerveDrive Odometry

abalakrishnan1 committed:
Merge remote-tracking branch 'origin/swerve-subsystem' into swerve-subsystem

# Conflicts:
#	.idea/jarRepositories.xml
#	src/main/java/frc/team3256/robot/subsystems/SwerveDrive.java

akhil-datla committed:
Moved constants of SwerveModule

abalakrishnan1 committed:
conflicts resolved

akhil-datla committed:
Update .gitignore with jarRepositories.xml

akhil-datla committed:
added SwerveDriveModule

viveknadig282 committed:
added controls

abalakrishnan1 committed:
changed code to match talonfx

abalakrishnan1 committed:
added values so build wouldn't fail

akhil-datla committed:
uncommented default swerve modules

viveknadig282 committed:
added Official SDS MK3 & MK4 Code

viveknadig282 committed:
Merge branch 'swerve-subsystem' of https://github.com/Team3256/Alpine into swerve-subsystem

viveknadig282 committed:
fixed merge errors in the official MK4 & MK3 code

abalakrishnan1 committed:
Added a class for the swerve constants

akhil-datla committed:
added Paths.java

adityapawar1 committed:
Create Auto Chooser and Add Chooser to Robot.java

Added paths and test trajectories to Paths.java. Also added the RobotContainer to select the selected command from SendableChooser.

StarDylan reviewed (commented):
Hey I know this is a work in progress, but I found some pretty minor things that might make the code a little more nice to work in. Let me know if some of my comments don't make sense or is just wrong. Anyway Nice Job!

viveknadig282 committed:
removed extra comments and changed code to match MK4

3LucasZ committed:
strange changes in config files

abalakrishnan1 committed:
commented swerve modules out for testing

akhil-datla committed:
change CAN ID constants

adityapawar1 committed:
Fixed Phoniex-latest.json to correct version

Downloaded latest working json, this fixed the robot code not deploying

adityapawar1 committed:
Finished adding offsets, testing all modules together

Finished setting motor ids using tuner and offsets while testing. Also uncommented all swerve modules in SwerveDrive.java

adityapawar1 committed:
Refoactor code to our convention

Remove m_ and moved constants to Constants.java

viveknadig282 committed:
Fixed some errors in spelling and removed extraneous code.

abalakrishnan1 committed:
Fixed the build failing due to the test cases

adityapawar1 committed:
Add logs for desired speed and fix offsets

Logging desired speeds and revised constant offsets

akhil-datla committed:
Add Field Control + Offsets

Field Oriented not working, Offsets needed to be negative

Co-authored-by: Dylan Starink <dylan@starink.com>
Co-authored-by: Aditya Pawar <34043950+adityapawar1@users.noreply.github.com>
Co-authored-by: vivek nadig <vivek.nadig@warriorlife.net>
Co-authored-by: Akhil Datla <66145155+akhil-datla@users.noreply.github.com>
Co-authored-by: Advay Balakrishnan <a.balakrishnan@warriorlife.net>

StarDylan committed:
Removed IDE Files

StarDylan committed:
Fixes Swerve ID Constants

Moved them from Swerve Constants to ID Constants,
Renamed motors and encoder IDs to have _ID at the end

StarDylan committed:
Cleans up Swerve Constants

Cleans up the constants so they are more
consistent and friendly to edit. SwerveDriveKinematics
is moved inside SwerveDrive.java since there is nothing
for technician to edit

StarDylan committed:
Fixes CANTest Build Fail

Moved CANConstants to IDConstants
and References Pigeon in IDConstants
As well as remove warning due to
`ids` always being the same constant

viveknadig282 committed:
changed gyro rotation to be positive

akhil-datla committed:
"Fix" field Oriented.

Co-authored-by: Dylan Starink <dylan@starink.com>
Co-authored-by: Aditya Pawar <34043950+adityapawar1@users.noreply.github.com>
Co-authored-by: vivek nadig <vivek.nadig@warriorlife.net>
Co-authored-by: Akhil Datla <66145155+akhil-datla@users.noreply.github.com>
Co-authored-by: Advay Balakrishnan <a.balakrishnan@warriorlife.net>

akhil-datla committed:
Merge remote-tracking branch 'origin/swerve-subsystem' into swerve-subsystem

# Conflicts:
#	src/main/java/frc/team3256/robot/Constants.java
#	src/main/java/frc/team3256/robot/auto/Paths.java
#	src/main/java/frc/team3256/robot/subsystems/SwerveDrive.java

adityapawar1 committed:
Fix auto chooser and add SwerveModuleState optimization

Used the WPILib SwerveModuleState.optimize to optmize module commands.
Added default do nothing option in AutoChooser and sent auto chooser to dashboard in Robot.java

abalakrishnan1 committed:
Added auto constants that were uninitialized

adityapawar1 committed:
Fix CANTest.java errors

abalakrishnan1 committed:
Merge remote-tracking branch 'origin/swerve-subsystem' into swerve-subsystem

# Conflicts:
#	src/main/java/frc/team3256/robot/helper/CANTest.java

adityapawar1 committed:
Fix CANTest.java errors

akhil-datla committed:
moved constants

akhil-datla committed:
switched to proper variable name convention

abalakrishnan1 committed:
Fixed auto constants

adityapawar1 committed:
Made constants default from WPILib

akhil-datla committed:
commented out unused constants

akhil-datla committed:
fixed convention

abalakrishnan1 committed:
Renamed the optimized swerve module states

abalakrishnan1 committed:
Merge remote-tracking branch 'origin/swerve-subsystem' into swerve-subsystem

adityapawar1 committed:
Reset pose everytime auto starts

adityapawar1 committed:
Merge remote-tracking branch 'origin/swerve-subsystem' into swerve-subsystem

abalakrishnan1 committed:
Changed var type to static (ProfiledPIDController) for consistency

abalakrishnan1 committed:
Got rid of auto constants that will never be used

akhil-datla committed:
Changed Gear to Fast, and added some PID Constants

akhil-datla committed:
Merge remote-tracking branch 'origin/swerve-subsystem' into swerve-subsystem

adityapawar1 committed:
Send trajctory to dashbaord in auto init

adityapawar1 committed:
Add trajectory following for swerve drive

Created TrajectoryFollowCommand.java to follow waypoints in a trajectory
Added feedforward constant for trajectory following
Created ProfiledPIDController for theta with a feedforwad
Calculate ChassisSpeeds based on odometry
Tuned PID controller for x translation and PIDF controller for theta

Co-authored-by: Aditya Pawar <aditya.pawar@warriorlife.net>
Co-authored-by: Advay Balakrishnan <a.balakrishnan@warriorlife.net>

abalakrishnan1 committed:
Merge remote-tracking branch 'origin/swerve-subsystem' into swerve-subsystem

adityapawar1 committed:
Theta setpoint reach desired angle at 3/4 through the trajectory

abalakrishnan1 committed:
Merge remote-tracking branch 'origin/swerve-subsystem' into swerve-subsystem

abalakrishnan1 committed:
Added UniformThetaSupplier for auto path
